src/lib/orionTypes/UpdateActionType.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+*
+* Copyright 2018 Telefonica Investigacion y Desarrollo, S.A.U
+*
+* This file is part of Orion Context Broker.
+*
+* Orion Context Broker is free software: you can redistribute it and/or
+* modify it under the terms of the GNU Affero General Public License as
+* published by the Free Software Foundation, either version 3 of the
+* License, or (at your option) any later version.
+*
+* Orion Context Broker is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Fermin Galan
+*/
+
+#include <string>
+#include <string.h>
+
+#include "common/globals.h"
+#include "orionTypes/UpdateActionType.h"
+
+
+
+/* ****************************************************************************
+*
+* actionTypeString -
+*/
+std::string actionTypeString(ApiVersion apiVersion, ActionType action)
```

Comment:
If this function could return a `char*` instead, we wouldn't have to copy an object onto the stack ...

Reply:
I tend to prefer returning std::string, for simplicity and readibility reasons (I think we have talked about this in the past... long time... many times :D)

Note that in the previous method equivalent to this (render() in the now deleted ngsi10/UpdateActionType.cpp file) was using the same pattern.

Reply:
Yeah ... you prefer C++
In reality, the win in simplicity and readibility you talk about doesn't really exist.
Just changing std::string for char* doesn't change simplicity and readibility at all.

The problem here is that we return am object on the stack and put own style guide forbids that.
[ Which is a thing all style guides do, for good reasons ]
I do know we return std::string in a million places. I just don't see why we have to add more functions returning objects. What we should do is to not add more and fis the old ones.

What about:

`void actionTypeString(ApiVersion apiVersion, ActionType action, std::string* typeStringP)` ?

That option keeps std::string and removes the problem.

Reply:
Comparing:

```
std::string actionTypeString(ApiVersion apiVersion, ActionType action) {...}
...
  std::string details = std::string("action: ") + actionTypeString(apiVersion, action) + ...
...
```

with 

```
void actionTypeString(ApiVersion apiVersion, ActionType action, std::string* typeStringP) {...}
...
  std::string actionTypeS;
  void actionTypeString(apiVersion, action, &typeStringP);
  std::string details = std::string("action: ") + actionTypeS + ...
...
```

The first option is more compact (less code and more "direct"), readable (less lines of code, less extra variables at the caller) and mantenible (less points to touch if a change is needed there).

However, I think the discussion is out of the scope of this PR, so I'd suggest keep as it was before my PR (note that the render() methods where this comes from already uses std::string to return) and discuss in it in an independent (and more general) scope.

(With regards to style, the rule currently in place has to be reviewed, as originally  `std::string` was allowed to be returned, as reference (`std::string&`)... before we realize `const` cannot be used here. Note the rule mentions "inefficient copies", which has to do with large objects. Is std::string a large object? Which size has a std::string? We should think in this kind of questions...)

Reply:
> Yeah ... you prefer C++

I don't have a *general* preference of C++ among C (or the other way around ;). However, *particularly* you are right I like std::string, as they are a very useful abstraction to deal with text strings.

Reply:
So, shall we follow our style guide, and avoid sending objects on the stack or should we change the style guide and say it's OK to send objects on the stack?

Never mind for this PR. It is true that we return std::string in many places already.
Creating an issue about removing all of them.

Reply:
NTC